core/src/main/java/org/pentaho/platform/scheduler2/action/SchedulerFilenameUtils.java

```diff
@@ -39,61 +38,126 @@ private SchedulerFilenameUtils() {
+   * @throws IllegalArgumentException if path contains windows separator defined by {@link #WINDOWS_PATH_SEPARATOR}
    */
   public static String concat( String basePath, String pathToAdd ) {
 
     if ( basePath == null || pathToAdd == null ) {
       return null;
     }
+
+    checkPaths( basePath, pathToAdd );
+
     /*
      * NOTE don't alter schema syntax ie remove // in a URI.
      * Don't call any function that eventually calls org.apache.commons.io.FilenameUtils#normalize
      * Similar logic to org.pentaho.platform.web.http.api.resources.SchedulerOutputPathResolver
      */
-    return getNoEndSeparator( basePath )  + PATH_SEPARATOR + FilenameUtils.getName( pathToAdd );
+    return getNoEndSeparator( toSb( basePath ) )
+        .append( PATH_SEPARATOR )
+        .append( getNoEndSeparator( toSb( pathToAdd ).reverse() ).reverse() ) // remove separator at the start of path
+        .toString();
   }
 
   /**
    * Return only the path without the ending separator defined by {@link #PATH_SEPARATOR}
    * @param path
    * @return <code>path</code> without the end separator
    */
-  protected static String getNoEndSeparator( String path ) {
-    return  ( StringUtils.isNotBlank( path )
-        && path.length() - 1 == FilenameUtils.indexOfLastSeparator( path ) )
-      ? path.substring( 0, path.length() - 1 ) // remove end separator
-      : path;
+  protected static StringBuilder getNoEndSeparator( StringBuilder path ) {
+    if ( path == null ) {
+      return null;
+    }
+
+    return ( indexOfLastSeparator( path ) != -1 && path.length() - 1 == indexOfLastSeparator( path )  )
+        ? path.deleteCharAt( path.length() - 1 ) // remove end separator
+        : path;
+  }
+
+  /**
+   * Return index of last separator defined by {@link #PATH_SEPARATOR}
+   * @param fileName
+   * @return 0th index of {@link #PATH_SEPARATOR}, -1 otherwise.
+   */
+  protected static int indexOfLastSeparator( StringBuilder fileName ) {
+    if ( fileName == null ) {
+      return -1;
+    } else {
+      return fileName.lastIndexOf( PATH_SEPARATOR );
+      /*
+       * NOTE: if we need to handle backward slash ie windows path separator
+       * see {@link org.apache.commons.io.FilenameUtils#indexOfLastSeparator}
+       */
+    }
+  }
+
+  /**
+   * Determines if all <code>paths</code> are valid.
+   * @param paths
+   * @throws IllegalArgumentException if a single path contains windows separator defined
+   * by {@link #WINDOWS_PATH_SEPARATOR}
+   */
+  protected static void checkPaths( String... paths ) {
+    for ( String path : paths ) {
+      checkPath( path );
+    }
+  }
+
+  /**
+   * Determines if <code>path</code> is valid.
+   * @param path
+   * @throws IllegalArgumentException if path contains windows separator defined by {@link #WINDOWS_PATH_SEPARATOR}
+   */
+  protected static void checkPath( String path ) {
+    if ( path.contains( WINDOWS_PATH_SEPARATOR ) ) {
+      throw new IllegalArgumentException(
```

Comment:
not a blocker, a question/thought on program flow: If multiple paths passed to checkPaths are illegal, will we see exceptions logged for all invalid paths?

Reply:
Good question.
No just the first illegal path will cause an` IllegalArgumentException` to be thrown and the function will exit with that error thrown. 

It's certainly possible to capture multiple paths, however, I wouldn't print the path due to possible security issues. So capturing multiple paths wouldn't give any more context. If a runtime error is thrown, I'm sure this exception would bubble up and someone looking at logs would the have necessary context to debug what the issue is.